test/integration/RedisResourceLockerIntegration.test.ts

```diff
@@ -0,0 +1,218 @@
+    const fileUrl = `${baseUrl}image.png`;
+    const fileData = await fs.readFile(
+      joinFilePath(__dirname, '../assets/testimage.png'),
+    );
+
+    let response = await fetch(fileUrl, {
+      method: 'PUT',
+      headers: {
+        'content-type': 'image/png',
+      },
+      body: fileData,
+    });
+    expect(response.status).toBe(205);
+
+    // GET
+    response = await fetch(fileUrl);
+    expect(response.status).toBe(200);
+    expect(response.headers.get('content-type')).toBe('image/png');
+    response = await fetch(fileUrl);
+    expect(response.status).toBe(200);
+    expect(response.headers.get('content-type')).toBe('image/png');
+    response = await fetch(fileUrl);
+    expect(response.status).toBe(200);
+    expect(response.headers.get('content-type')).toBe('image/png');
+    response = await fetch(fileUrl);
+    expect(response.status).toBe(200);
+    expect(response.headers.get('content-type')).toBe('image/png');
+
+    // DELETE
+    response = await fetch(fileUrl, {
+      method: 'DELETE',
+    });
+    expect(response.status).toBe(205);
+    response = await fetch(fileUrl);
+    expect(response.status).toBe(404);
+  });
+
+  describe('Test the ResourceLocker itself', (): void => {
+    const identifier = { path: 'http://test.com/foo' };
+
+    it('can lock and unlock a resource.', async(): Promise<void> => {
+      await expect(locker.acquire(identifier)).resolves.toBeUndefined();
+      await expect(locker.release(identifier)).resolves.toBeUndefined();
+    });
+
+    it('can lock a resource again after it was unlocked.', async(): Promise<void> => {
+      await expect(locker.acquire(identifier)).resolves.toBeUndefined();
+      await expect(locker.release(identifier)).resolves.toBeUndefined();
+      await expect(locker.acquire(identifier)).resolves.toBeUndefined();
+      await expect(locker.release(identifier)).resolves.toBeUndefined();
+    });
+
+    it('can acquire different keys simultaneously.', async(): Promise<void> => {
+      const lock1 = locker.acquire({ path: 'path1' });
+      const lock2 = locker.acquire({ path: 'path2' });
+      const lock3 = locker.acquire({ path: 'path3' });
+
+      await expect(Promise.all([ lock1, lock2, lock3 ])).resolves.toBeDefined();
+
+      await locker.release({ path: 'path1' });
+      await locker.release({ path: 'path2' });
+      await locker.release({ path: 'path3' });
+    });
+
+    it('Cannot acquire the same lock simultaneously.', async(): Promise<void> => {
+      let res = '';
+      const lock1 = locker.acquire(identifier);
+      const lock2 = locker.acquire(identifier);
+      const lock3 = locker.acquire(identifier);
+
+      const l1 = lock1.then(async(): Promise<void> => {
+        res += 'l1';
+        await locker.release(identifier);
+        res += 'r1';
+      });
+      const l2 = lock2.then(async(): Promise<void> => {
+        res += 'l2';
+        await locker.release(identifier);
+        res += 'r2';
+      });
```

Comment:
```suggestion
      await new Promise((resolve): any => setImmediate(resolve));

      const l2 = lock2.then(async(): Promise<void> => {
        res += 'l2';
        await locker.release(identifier);
        res += 'r2';
      });
      const l1 = lock1.then(async(): Promise<void> => {
        res += 'l1';
        await locker.release(identifier);
        res += 'r1';
      });
```

Just to really make sure the test doesn't currently succeed due to the order or redlock not having sent a request yet.